src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Windows/CertificatePal.PrivateKey.cs

```diff
@@ -239,5 +339,223 @@ private CspParameters GetPrivateKeyCsp()
+            bool machineKey = cngKey.IsMachineKey;
+
+            // CAPI RSA_SIGN keys won't open correctly under CNG without the key number being specified, so
+            // check to see if we can figure out what key number it needs to re-open.
+            int keySpec = GuessKeySpec(provider, keyName, machineKey, cngKey.AlgorithmGroup);
+
+            CRYPT_KEY_PROV_INFO keyProvInfo = new CRYPT_KEY_PROV_INFO();
+
+            fixed (char* keyNamePtr = cngKey.KeyName)
+            fixed (char* provNamePtr = cngKey.Provider.Provider)
+            {
+                keyProvInfo.pwszContainerName = keyNamePtr;
+                keyProvInfo.pwszProvName = provNamePtr;
+                keyProvInfo.dwFlags = machineKey ? CryptAcquireContextFlags.CRYPT_MACHINE_KEYSET : 0;
+                keyProvInfo.dwKeySpec = keySpec;
+
+                if (!Interop.crypt32.CertSetCertificateContextProperty(
+                    pal._certContext,
+                    CertContextPropId.CERT_KEY_PROV_INFO_PROP_ID,
+                    CertSetPropertyFlags.None,
+                    &keyProvInfo))
+                {
+                    pal.Dispose();
+                    throw Marshal.GetLastWin32Error().ToCryptographicException();
+                }
+            }
+
+            return pal;
+        }
+
+        private static int GuessKeySpec(
+            CngProvider provider,
+            string keyName,
+            bool machineKey,
+            CngAlgorithmGroup algorithmGroup)
+        {
+            if (provider == CngProvider.MicrosoftSoftwareKeyStorageProvider ||
+                provider == CngProvider.MicrosoftSmartCardKeyStorageProvider)
+            {
+                // Well-known CNG providers, keySpec is 0.
+                return 0;
+            }
+
+            const int NTE_BAD_KEYSET = unchecked((int)0x80090016);
+
+            try
+            {
+                CngKeyOpenOptions options = machineKey ? CngKeyOpenOptions.MachineKey : CngKeyOpenOptions.None;
+
+                using (CngKey.Open(keyName, provider, options))
+                {
+                    // It opened with keySpec 0, so use keySpec 0.
+                    return 0;
+                }
+            }
+            catch (CryptographicException e)
+            {
+                Debug.Assert(
+                    e.HResult == NTE_BAD_KEYSET,
+                    $"CngKey.Open had unexpected error: 0x{e.HResult:X8}: {e.Message}");
+
+                CspParameters cspParameters = new CspParameters
+                {
+                    ProviderName = provider.Provider,
+                    KeyContainerName = keyName,
+                    Flags = CspProviderFlags.UseExistingKey,
+                    KeyNumber = (int)KeyNumber.Signature,
+                };
+
+                if (machineKey)
+                {
+                    cspParameters.Flags |= CspProviderFlags.UseMachineKeyStore;
+                }
+
+                if (algorithmGroup == CngAlgorithmGroup.Rsa)
```

Comment:
The important part of this being that since we see it as a CNG key, we can't ask if it was AT_EXCHANGE or AT_SIGNATURE; but on downlevel OSes CNG only tried AT_EXCHANGE (if not told explicitly); so we need to sniff out AT_SIGNATURE keys.